### PR TITLE
feat: New IO descriptor

### PR DESCRIPTION
**Pull request by frostming**
Signed-off-by: Frost Ming <me@frostming.com>

## What does this PR address?

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/bentoml/_pydantic.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+import typing as t
+
+from pydantic._internal import _known_annotated_metadata
+from typing_extensions import get_args
+from typing_extensions import get_origin
```
Comment (aarnphm) [inline src/bentoml/_pydantic.py:7]: Should we have a `bentoml._typing_compat`?

File: setup.py [hunk 2]
```diff
@@ -1,27 +0,0 @@
-import shutil
```
Comment (aarnphm) [inline setup.py:1]: we still need to keep this file so that github can detect the pypi repo

File: src/bentoml/_internal/ionext/client.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+                input=route["input"],
+                output=route["output"],
+                doc=route.get("doc"),
+                stream_output=route["output"].get("is_stream", False),
+            )
+
+        self.__attrs_init__(url=url, endpoints=routes, media_type=media_type)
+        self.__post_init__()
+
+    def __post_init__(self) -> None:
```
Comment (aarnphm) [inline src/bentoml/_internal/ionext/client.py:87]: Shouldn't we just use `__attrs_post_init__`?

File: setup.py (hunk 2 above)
Comment (aarnphm) [reply setup.py:1]: idk if the github dependency graph support pyproject.toml

File: src/bentoml_io/serde.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return pickle.dumps(model_data)
+
+    def deserialize_model(self, model_bytes: bytes, cls: type[T]) -> T:
+        obj = pickle.loads(model_bytes)
+        if not isinstance(obj, cls):
+            obj = cls.model_validate(obj)
+        return obj
+
+    def serialize(self, obj: t.Any) -> bytes:
+        return pickle.dumps(obj)
```
Comment (aarnphm) [inline src/bentoml_io/serde.py:66]: https://peps.python.org/pep-0574/

worth to also implement something like `bentoml._internal.utils.pickle.pep574_dumps/pep574_loads` for `PickleSerde.serialize/deserialize`

File: setup.py (hunk 2 above)
Comment (frostming) [reply setup.py:1]: Yes it does

File: src/bentoml_io/models.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            annotation = field.annotation
+            if is_list_type(annotation):
+                annotation = get_args(annotation)[0]
+            if is_annotated(annotation):
+                annotation = get_args(annotation)[0]
+            if is_file_like(annotation) or is_image_type(annotation):
+                cls.multipart_fields.append(k)
+
+    @classmethod
+    async def from_http_request(cls, request: Request, serde: Serde) -> IODescriptor:
```
Comment (aarnphm) [inline src/bentoml_io/models.py:44]:
```suggestion
    async def from_http(cls, request: Request, serde: Serde) -> IODescriptor:
```
Should we do `from_http` and `to_http`?

File: src/bentoml/_internal/marshal/dispatcher.py [hunk 6]
```diff
@@ -18,8 +18,7 @@
 
 
 if t.TYPE_CHECKING:
-    from ..runner.container import Payload
-    from ..runner.utils import Params
+    pass
```
Comment (aarnphm) [inline src/bentoml/_internal/marshal/dispatcher.py:21]: Should we remove this branch then?

File: src/bentoml_io/models.py (hunk 5 above)
Comment (frostming) [reply src/bentoml_io/models.py:44]: `from_http` doesn't tell if it's from request or response

File: src/bentoml_io/client/http.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+
+    async def close(self) -> None:
+        if self._client is not None and not self._client.closed:
+            await self._client.close()
+
+    async def __aexit__(self, *args: t.Any) -> None:
+        return await self.close()
+
+
+class SyncHTTPClient(HTTPClient):
```
Comment (aarnphm) [inline src/bentoml_io/client/http.py:347]: wouldn't this class will have `__aenter__` and `__aexit__`?

Comment (aarnphm) [reply src/bentoml_io/client/http.py:347]: We should raise error immediately if they try to do async with?

File: src/bentoml_io/models.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    from .serde import Serde
+
+
+DEFAULT_STREAM_MEDIA_TYPE = "text/event-stream"
+DEFAULT_TEXT_MEDIA_TYPE = "text/plain"
+
+
+class IOMixin:
+    multipart_fields: ClassVar[list[str]]
+    media_type: ClassVar[str | None] = None
```
Comment (aarnphm) [inline src/bentoml_io/models.py:32]:
```suggestion
    multipart_fields: ClassVar[t.List[str]]
    media_type: ClassVar[t.Optional[str]] = None
```

Screenshot 2023-11-08 at 03 32 08

We need to be careful with this syntax for serialisation purposes 🤔

File: src/bentoml_io/servable.py [hunk 9]
```diff
@@ -36,11 +37,16 @@ def async_client(self, name_or_class: str | type[Servable]) -> AbstractClient:
 
     @classmethod
     def schema(cls) -> dict[str, t.Any]:
-        return {
-            "name": cls.name,
-            "type": "service",
-            "routes": [method.schema() for method in cls.__servable_methods__.values()],
-        }
+        return dict_filter_none(
+            {
```
Comment (aarnphm) [inline src/bentoml_io/servable.py:41]: Should make a `TypedDict` for this? Same with API

File: src/bentoml_io/arrow.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    with arrow_serialization():
+        table = pa.Table.from_pylist([model.model_dump()], schema=arrow_schema)
+    with pa.ipc.new_stream(out_stream, arrow_schema) as writer:
+        writer.write_table(table)
+
+
+def deserialize_from_arrow(model: type[T], in_stream: t.BinaryIO) -> T:
+    with pa.ipc.open_stream(in_stream) as reader:
+        df = reader.read_pandas()
+        ins = df.to_dict(orient="records")[0]
```
Comment (judahrand) [inline src/bentoml_io/arrow.py:103]: This is bonkers way of doing this. Why not:

```python
        ins = reader.read_all().to_pylist()[0]
```

Comment (frostming) [reply src/bentoml_io/arrow.py:103]: Would you like to contribute?

**Pull request merged**